meilisearch-index-setting-macro/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+    attrs: &[syn::Attribute],
+    attribute_count: &mut std::collections::HashMap<String, i32>,
+) -> std::result::Result<Vec<String>, proc_macro2::TokenStream> {
+    let mut value: Vec<String> = vec![];
+    for attr in attrs {
+        if let std::result::Result::Ok(syn::Meta::List(list)) = attr.parse_meta() {
+            if list.path.is_ident("document") {
+                for nested_meta in list.nested {
+                    if let syn::NestedMeta::Meta(syn::Meta::Path(path)) = nested_meta {
+                        for segment in path.segments {
+                            value.push(segment.ident.to_string());
+                            *attribute_count
+                                .entry(segment.ident.to_string())
+                                .or_insert(0) += 1;
+
+                            if segment.ident == "primary_key"
+                                && attribute_count.get("primary_key").unwrap_or(&0) > &1
+                            {
+                                return std::result::Result::Err(
+                                    syn::Error::new(
+                                        segment.ident.span(),
+                                        "primary_key already exists",
+                                    )
+                                    .to_compile_error(),
+                                );
+                            }
+                            if segment.ident == "distinct"
+                                && attribute_count.get("distinct").unwrap_or(&0) > &1
+                            {
+                                return std::result::Result::Err(
+                                    syn::Error::new(
+                                        segment.ident.span(),
+                                        "distinct already exists",
+                                    )
+                                    .to_compile_error(),
+                                );
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    std::result::Result::Ok(value)
+}
+
+fn get_settings_token_for_list(
+    attributes: &Vec<String>,
+    attribute_key: &str,
+) -> proc_macro2::TokenStream {
+    let string_attributes = attributes.iter().map(|attr| {
+        quote! {
+            #attr
+        }
+    });
+
+    let ident = syn::Ident::new(attribute_key, proc_macro2::Span::call_site());
+
+    if !attributes.is_empty() {
+        quote! {
+            .#ident([#(#string_attributes),*])
+        }
+    } else {
+        proc_macro2::TokenStream::new()
+    }
+}
+
+fn get_settings_token_for_string(
+    attribute: &String,
+    attribute_key: &str,
+) -> proc_macro2::TokenStream {
+    let ident = syn::Ident::new(attribute_key, proc_macro2::Span::call_site());
+
+    if !attribute.is_empty() {
+        quote! {
+            .#ident(#attribute)
+        }
+    } else {
+        proc_macro2::TokenStream::new()
```

Comment:
With the same idea as for vec attributes, we should be able to provide `None`. Unfortunately this is not possible since `with_distinct_attribute` does not accept a `None` value.

I'll open an issue about it

Reply:
https://github.com/meilisearch/meilisearch-rust/issues/368